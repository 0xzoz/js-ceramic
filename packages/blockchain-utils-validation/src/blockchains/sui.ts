import { BlockchainHandler } from '../blockchain-handler.js'
import { LinkProof } from '@ceramicnetwork/blockchain-utils-linking'
import * as uint8arrays from 'uint8arrays'
import { normalizeAccountId } from '@ceramicnetwork/common'
import nacl from 'tweetnacl';
import { Base64DataBuffer } from "@mysten/sui.js";

const verifySignature = async (
  pubKey: Uint8Array,
  message: string,
  signature: Uint8Array
): Promise<boolean> => {
  const signData = new Base64DataBuffer(new TextEncoder().encode(message));
  const verified = nacl.sign.detached.verify(
    signData.getData(),
    signature,
    pubKey
  )
  return verified
}

const namespace = 'sui'

export async function validateLink(proof: LinkProof): Promise<LinkProof | null> {
  const pubKey = uint8arrays.fromString(normalizeAccountId(proof.account).address, 'base64')
  const msg = proof.message
  const sig = uint8arrays.fromString(proof.signature, 'base64')
  const is_sig_valid = await verifySignature(pubKey, msg, sig)
  return is_sig_valid ? proof : null
}

export const handler: BlockchainHandler = {
  namespace,
  validateLink,
}
