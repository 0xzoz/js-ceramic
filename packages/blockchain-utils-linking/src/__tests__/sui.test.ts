import { jest } from '@jest/globals'
import { SuiAuthProvider } from '../sui.js'
import { Base64DataBuffer,Ed25519Keypair } from "@mysten/sui.js";


const did = 'did:3:bafysdfwefwe'
const privKey = 'mdqVWeFekT7pqy5T49+tV12jO0m+ESW7ki4zSU9JiCgbL0kJbj5dvQ/PqcDAzZLZqzshVEs01d1KZdmLh4uZIg=='
const chainRef = 'mainnet'

class MyWalletAdapter  {
  readonly _keyPair: Ed25519Keypair

  constructor(keyPair: Ed25519Keypair) {
    this._keyPair = keyPair
  }

  async signMessage(message: Uint8Array): Promise<Uint8Array> {
    const msg = new Base64DataBuffer(message);
    const signature = this._keyPair.signData(msg);
    return signature.getData()
  }
}

let keyPairEd25519: Ed25519Keypair

beforeAll(() => {
  keyPairEd25519 = Ed25519Keypair.fromSecretKey(Buffer.from(privKey, 'base64'))
  global.Date.now = jest.fn().mockImplementation(() => 666000)
})

afterAll(() => {
  jest.clearAllMocks()
})

describe('Blockchain: Sui', () => {
  describe('createLink', () => {
    test(`create proof for ${chainRef}`, async () => {
      const keypair = keyPairEd25519
      const provider = new MyWalletAdapter(keypair)
      const authProvider = new SuiAuthProvider(
        provider,
        keyPairEd25519.getPublicKey().toString(),
        chainRef
      )
      const proof = await authProvider.createLink(did)
      expect(proof).toMatchSnapshot()
    })
  })

  describe('authenticate', () => {
    test(`create proof for ${chainRef}`, async () => {
      const provider = new MyWalletAdapter(keyPairEd25519)
      const authProvider = new SuiAuthProvider(
        provider,
        keyPairEd25519.getPublicKey().toString(),
        chainRef
      )
      const result = await authProvider.authenticate('msg')
      expect(result).toMatchSnapshot()
    })
  })
})