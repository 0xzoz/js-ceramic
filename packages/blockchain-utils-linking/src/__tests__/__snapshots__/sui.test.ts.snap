// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blockchain: Sui authenticate create proof for mainnet 1`] = `"0xd042f28e5ff37a2f68f7d7b96adca0d10b5cba3d4d6d056796653479d8a70c0e"`;

exports[`Blockchain: Sui createLink create proof for mainnet 1`] = `
Object {
  "account": "Gy9JCW4+Xb0Pz6nAwM2S2as7IVRLNNXdSmXZi4eLmSI=@sui:mainnet",
  "message": "Link this account to your identity

did:3:bafysdfwefwe 
Timestamp: 666",
  "signature": "KriuHY8KAU87qNeGjdtvutl0TWgSnOPpUDI8fniN8jHGdwcKubYUooJja/hnFZoizq7q66dx5XRPxKS4EcnGDA",
  "timestamp": 666,
  "type": "sui",
  "version": 2,
}
`;
